### PR TITLE
Use environment variables in config files #60

### PR DESCRIPTION
**Pull request by navinrathore**
Feature to allow user to pass environment variables for the Zingg configuration values #60

### CODE REVIEW

**Comment by sonalgoyal** [comment]
Junis are meant to be atomic. So you build tests like testBooleanVar teststringVar testIntegerVar testDataPipeFormat testOutputPipexxxx testDataPipeArrayFormat etc etc...go as many cases as you can think of, and combinations testBooleanStringxxxxx 

the tests also need to address the specific issue - values of pipe properties

tests should also be there for malformed system var like $var or var$ - in case of boolean and int you should get exceptions, in other cases you should get exceptions again

and tests when something is referenced in json but not configured in env variable also need to be there.

**Comment by sonalgoyal** [comment]
Also checkin comments need to be better - right now there are three commits with Use environment variables in config files. You can reference the issue in all commits, but the commit has to specify what is being changed too.

**Comment by navinrathore** [comment]
The newly added testcases cover the above suggestions and other units/fragments of the  the code changes. I tried to avoid redundant and out of scope scenarios.

File: client/src/main/java/zingg/client/Arguments.java [hunk 1]
```diff
@@ -189,7 +199,46 @@ public static final Arguments createArgumentsFromJSONString(String data, String
 					". The error is " + e.getMessage());
 		}
 	}
-	
+
+	public static final Arguments createArgumentsFromJSONTemplate(String filePath, String phase)
+			throws ZinggClientException {
```
Comment (navinrathore) [inline client/src/main/java/zingg/client/Arguments.java:204]: New Function added. 
Veryfing the code comments blocks added or not!

**Review by sonalgoyal — CHANGES_REQUESTED** [review]

File: client/src/main/java/zingg/client/Arguments.java (hunk 1 above)
Comment (navinrathore) [reply client/src/main/java/zingg/client/Arguments.java:204]: Disregarded

File: client/src/test/java/zingg/client/TestArguments.java [hunk 2]
```diff
@@ -54,7 +92,8 @@ public void testCreateArgsMissingMatchFile() {
 	}
 
 	@Test
-	public void testCreateArgsMissingDelimiterFile() {
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply client/src/test/java/zingg/client/TestArguments.java:57]: done

File: client/src/main/java/zingg/client/Arguments.java [hunk 3]
```diff
@@ -189,7 +199,46 @@ public static final Arguments createArgumentsFromJSONString(String data, String
...
+		try {
+			LOG.warn("Config Argument is " + filePath);
+			byte[] encoded = Files.readAllBytes(Paths.get(filePath));
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply client/src/main/java/zingg/client/Arguments.java:207]: It is a Java 11 function. No change is done.

File: client/src/main/java/zingg/client/Arguments.java [hunk 4]
```diff
@@ -189,7 +199,46 @@ public static final Arguments createArgumentsFromJSONString(String data, String
...
+			if (variables.containsKey(matcher.group(1))) {
+				String replacement = variables.get(matcher.group(1));
+				if (replacement == null || replacement.equals("")) {
+					LOG.warn("The environment variable for {" + matcher.group(1) + "} is not set or is empty string");
+				}
+				// quote to work properly with $ and {,} signs
+				matcher.appendReplacement(buffer, replacement != null ? Matcher.quoteReplacement(replacement) : "null");
+			}
+			else {
+				LOG.warn("The environment variable for {" + matcher.group(1) + "} is not set");
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply client/src/main/java/zingg/client/Arguments.java:235]: Added Exception.

**Comment by navinrathore** [comment]
Made changes as per comments and the discussion

**Review by sonalgoyal — CHANGES_REQUESTED** [review]
Overall it is much better. few things 
- print at warn level which variables are getting substituted - only name, not value.
- test that $var without an end $ throws an exception
-  is there a way to ensure or test non .env file has no substitutions.

File: client/src/test/java/zingg/client/TestArguments.java [hunk 5]
```diff
@@ -1,136 +1,202 @@
...
+			byte[] encoded = Files
+					.readAllBytes(Paths.get(getClass().getResource("../../testConfigTemplate.json.env").getFile()));
+
+			String template = new String(encoded, StandardCharsets.UTF_8);
+			String json = Arguments.substituteVariables(template, env);
+			Arguments args = Arguments.createArgumentsFromJSONString(json, "");
+
+			fail("Exception was expected for blank value for an environment variable");
+ 		} catch (IOException | ZinggClientException e) {
+ 			System.out.println("Expected exception received due to blank value for an environment variable");
```
[the remark this answers is not in the document]
Comment (navinrathore) [reply client/src/test/java/zingg/client/TestArguments.java:78]: ok

**Comment by navinrathore** [comment]
Made changes to add Logger in Tests; Added message for substituted environment variables as well.

**Pull request merged**